ethereum/executionclient/src/main/java/tech/pegasys/teku/ethereum/executionclient/schema/Response.java

```diff
@@ -41,15 +45,39 @@ public static <T> Response<T> withErrorMessage(final String errorMessage) {
     return new Response<>(null, errorMessage);
   }
 
-  public static <T, R> Response<R> unwrap(
-      final Response<T> response, final Function<T, R> unwrapFunction) {
+  public static <T, R> Response<R> unwrapVersioned(
+      final Response<T> response,
+      final Function<T, R> unwrapFunction,
+      final SpecMilestone expectedMilestone,
+      final Function<T, SpecMilestone> unwrapVersionFunction,
+      final boolean strictVersionCheck) {
+
     if (response.isFailure()) {
       return Response.withErrorMessage(response.getErrorMessage());
     }
     final T payload = response.getPayload();
-    return payload == null
-        ? Response.withNullPayload()
-        : new Response<>(unwrapFunction.apply(payload));
+    if (payload == null) {
+      return Response.withNullPayload();
+    }
+
+    final SpecMilestone receivedMilestone = unwrapVersionFunction.apply(payload);
+
+    final boolean milestonesAreMatching = receivedMilestone.equals(expectedMilestone);
+
+    if (strictVersionCheck && !milestonesAreMatching) {
+      throw new IllegalArgumentException(
+          "Wrong response version: expected "
+              + expectedMilestone
+              + ", received "
+              + receivedMilestone);
+    } else {
+      LOG.warn(
```

Comment:
Isn't this always going to log warning, regardless if milestones are matching or not for `getPayload`?

Reply:
🤦